### PR TITLE
Add OracleManagedDataAccess Aspire Component

### PR DESCRIPTION
**Pull request by andrevlins**
Adds the ability to connect to an Oracle database through the OracleManagedDataAccess library by adding a connection of type OracleConnection or DbConnection.

It also adds the ability to configure a container with an Oracle Database Free image (container-registry.oracle.com/database/free:latest).

To continue this work I need the following packages to be available in the nuget repository of this project:

AspNetCore.HealthChecks.Oracle 7.0.0 
Oracle.ManagedDataAccess.Core 23.3.0-dev 
Oracle.ManagedDataAccess.OpenTelemetry 23.2.0-dev

### CODE REVIEW

**Comment by andrevlins** [comment]
Hello @RussKie, Can you help me with these packages mentioned above?

AspNetCore.HealthChecks.Oracle 7.0.0
Oracle.ManagedDataAccess.Core 23.3.0-dev
Oracle.ManagedDataAccess.OpenTelemetry 23.2.0-dev

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    private static void RegisterOracleManagedDataAccessServices(this IHostApplicationBuilder builder, OracleManagedDataAccessSettings settings, string configurationSectionName, string connectionName, object? serviceKey)
+    {
+        if (serviceKey is null)
+        {
+            // delay validating the ConnectionString until the DataSource is requested. This ensures an exception doesn't happen until a Logger is established.
+            builder.Services.AddTransient((serviceProvider) =>
+            {
+                ValidateConnection();
+
+                return new OracleConnection(settings.ConnectionString);
```
Comment (davidfowl) [inline src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:94]: Oracle doesn't have a data source?

Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:94]: No. I intend to update this code as soon as they implement DbDataSource.

Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:94]: FYI 

oracle/dotnet-db-samples#329

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    private const string DefaultConfigSectionName = "Aspire:OracleManagedDataAccess";
+
+    /// <summary>
+    /// Registers <see cref="OracleConnection"/> for connecting Oracle database with Oracle.ManagedDataAccess client.
+    /// Configures health check, logging and telemetry for the Oracle.ManagedDataAccess client.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHostApplicationBuilder" /> to read config from and add services to.</param>
+    /// <param name="connectionName">A name used to retrieve the connection string from the ConnectionStrings configuration section.</param>
+    /// <param name="configureSettings">An optional delegate that can be used for customizing options. It's invoked after the settings are read from the configuration.</param>
+    /// <remarks>Reads the configuration from "Aspire:Oracle" section.</remarks>
```
Comment (mitchdenny) [inline src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:28]: Aspire:OracleManagedDataAccess

File: tests/Aspire.OracleManagedDataAccess.Tests/AspireOracleManagedDataAccessExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Oracle.ManagedDataAccess.Client;
+using Xunit;
+
+namespace Aspire.OracleManagedDataAccess.Tests;
+
+public class AspireOracleManagedDataAccessExtensionsTests
+{
+    private const string ConnectionString = "user id=system;password=123;data source=localhost:1522/freepdb1";
```
Comment (mitchdenny) [inline tests/Aspire.OracleManagedDataAccess.Tests/AspireOracleManagedDataAccessExtensionsTests.cs:14]: Is this a well known connection string?

File: Directory.Packages.props [hunk 4]
```diff
@@ -74,6 +74,7 @@
     <PackageVersion Include="MySqlConnector.DependencyInjection" Version="2.3.1" />
     <PackageVersion Include="Npgsql.DependencyInjection" Version="8.0.0-rc.2" />
     <PackageVersion Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="8.0.0-rc.2" />
+    <PackageVersion Include="Oracle.ManagedDataAccess.Core" Version="23.3.0-dev" />
```
Comment (eerhardt) [inline Directory.Packages.props:77]: We should be depending on stable versions, if possible. Is a stable version going to ship soon?

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+using Oracle.ManagedDataAccess.Client;
+
+namespace Microsoft.Extensions.Hosting;
+
+/// <summary>
+/// Extension methods for connecting Oracle database with Oracle.ManagedDataAccess client
+/// </summary>
+public static class AspireOracleManagedDataAccessExtensions
+{
+    private const string DefaultConfigSectionName = "Aspire:OracleManagedDataAccess";
```
Comment (eerhardt) [inline src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:19]:
```suggestion
    private const string DefaultConfigSectionName = "Aspire:Oracle:ManagedDataAccess:Core";
```

This should follow the name of the component.

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                timeout: default));
+        }
+    }
+
+    private static void RegisterOracleManagedDataAccessServices(this IHostApplicationBuilder builder, OracleManagedDataAccessSettings settings, string configurationSectionName, string connectionName, object? serviceKey)
+    {
+        if (serviceKey is null)
+        {
+            // delay validating the ConnectionString until the DataSource is requested. This ensures an exception doesn't happen until a Logger is established.
+            builder.Services.AddTransient((serviceProvider) =>
```
Comment (eerhardt) [inline src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs:90]: The connection should be transient? I would have guessed scoped.

File: src/Components/Aspire.OracleManagedDataAccess/OracleManagedDataAccessSettings.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /// <para>Gets or sets a boolean value that indicates whether the Open Telemetry tracing is enabled or not.</para>
+    /// <para>Enabled by default.</para>
+    /// </summary>
+    public bool Tracing { get; set; } = true;
+
+    /// <summary>
+    /// <para>Gets or sets a boolean value that indicates whether the Open Telemetry metrics are enabled or not.</para>
+    /// <para>Enabled by default.</para>
+    /// </summary>
+    public bool Metrics { get; set; } = true;
```
Comment (eerhardt) [inline src/Components/Aspire.OracleManagedDataAccess/OracleManagedDataAccessSettings.cs:32]: Are these used? If not, we shouldn't have properties for them.

File: src/Components/Aspire.OracleManagedDataAccess/ConfigurationSchema.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{
+  "properties": {
```
Comment (eerhardt) [inline src/Components/Aspire.OracleManagedDataAccess/ConfigurationSchema.json:2]: Is there integration with Logging? If so, can we add the log categories to the ConfigurationSchema.json?

See https://github.com/dotnet/aspire/pull/430 for example.

**Comment by mitchdenny** [comment]
@andrevlins did you mean to close this?

**Comment by smitpatel** [comment]
Force push erased all the commits which were here so github closed the PR. 😟

**Comment by andrevlins** [comment]
@mitchdenny I'll reopen with the new folder structure

File: src/Components/Aspire.OracleManagedDataAccess/OracleManagedDataAccessSettings.cs (hunk 7 above)
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/OracleManagedDataAccessSettings.cs]: Removed for now, I looked in the ODP.NET documentation and there are still no metrics that we can use here.

File: src/Components/Aspire.OracleManagedDataAccess/ConfigurationSchema.json (hunk 8 above)
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/ConfigurationSchema.json]: Not for the time being. I believe that if they implement DbDataSource they will provide this feature.

FYI: oracle/dotnet-db-samples#329

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs (hunk 5 above)
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs]: Sure

File: Directory.Packages.props (hunk 4 above)
Comment (andrevlins) [reply Directory.Packages.props:78]: This is the version that supports Open Telemetry. My intention is to evolve this PR as they develop the stable version.

File: src/Components/Aspire.OracleManagedDataAccess/OracleHealthCheck.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Oracle.ManagedDataAccess.Client;
+
+namespace Aspire.OracleManagedDataAccess;
+internal sealed class OracleHealthCheck : IHealthCheck
```
[the remark this answers is not in the document]
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/OracleHealthCheck.cs]: Sure

File: src/Components/Aspire.OracleManagedDataAccess/Aspire.OracleManagedDataAccess.csproj [hunk 10]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/Aspire.OracleManagedDataAccess.csproj]: Done. I ended up closing the PR unintentionally during this process, sorry about that

File: src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs (hunk 2 above)
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/AspireOracleManagedDataAccessExtensions.cs]: Changed to Aspire:Oracle:ManagedDataAccess:Core as @eerhardt advised

File: src/Components/Aspire.OracleManagedDataAccess/Aspire.OracleManagedDataAccess.csproj (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrevlins) [reply src/Components/Aspire.OracleManagedDataAccess/Aspire.OracleManagedDataAccess.csproj]: Done

File: Directory.Packages.props (hunk 4 above)
Comment (eerhardt) [reply Directory.Packages.props:78]: Perfect. That makes sense to me.

**Comment by eerhardt** [comment]
@andrevlins - this PR is marked as a Draft. Is it ready for review?

**Comment by andrevlins** [comment]
> @andrevlins - this PR is marked as a Draft. Is it ready for review?

This PR depends on a version of the Oracle.ManagedDataAccess nuget package that is still in preview.
The main reason is that it is the version that supports Open Telemetry.

https://github.com/dotnet/aspire/pull/1004#discussion_r1408649387

**Comment by andrevlins** [comment]
> do you have a timeline for when you think this PR will be ready for review?

I am waiting for the stable version of the library ```Oracle.ManagedDataAccess.Core``` and ```Oracle.ManagedDataAccess.OpenTelemetry``` version 23.x.x due to OpenTelemetry support.

Some paths we can follow:

- Continue with the PR in draft until the release of version 23.x.x.
- Use the ```dev```` version and post the PR for review (I don't think working with unstable versions is viable in this project).
- Downgrade the version to 3.21.xxx and not support OpenTelemetry (leaving this for another PR once the version is released).
- Close the PR so as not to disturb your management and open it again when the version is released.

What do you think?

**Comment by eerhardt** [comment]
I'm inclined to say that if no progress is being made on this PR, let's close it for now. We can always re-open it when progress is ready to be made.

**Pull request closed without merging**